camera_models.py

```diff
@@ -644,4 +645,152 @@ def point_3d_2_pixel(self, point_3d, normalized=False):
+                                      [ 0,      0      ,                1.0]]).to(dtype=uv.dtype, device=uv.device)
+
+        xyz = inv_intrinsics @ uv1
+
+        # Normalize rays to be unit length.
+        xyz = xyz / torch.linalg.norm(xyz, dim = -2, keepdims= True)
+
+        # Mask points that are out of field of view.
+        # NOTE(yoraish): why should there be any??
+        if len(uv.shape) == 2:
+            mask = torch.ones(uv.shape[1])
+        if len(uv.shape) == 3:
+            mask = torch.ones((uv.shape[0], uv.shape[2]))
+        return xyz, mask
+        
+
+    def point_3d_2_pixel(self, point_3d, normalized=False):
+        '''
+        Arguments:
+        point_3d (Tensor): A 3xN Tensor contains 3D point coordinates. 
+        normalized (bool): If True, then the returned coordinates are normalized to [-1, 1]
+        
+        NOTE: point_3d can also have a dimension of Bx3xN, where B is the 
+        batch number.
+        
+        Returns: 
+        A 2xN Tensor representing the 2D pixels. Bx2xN if batched.
+        A (N,) Tensor representing the valid mask. BxN if batched.
+        '''
+
+        point_3d = self.in_wrap(point_3d)
+
+        # Projection matrix.
+        intrinsics = torch.tensor([[self.fx, 0      , self.cx ],
+                                   [ 0,      self.fy, self.cy ],
+                                   [ 0,      0      ,      1.0]]).to(dtype= point_3d.dtype, device= point_3d.device)
+
+        # Pixel coordinates. 
+        uv_unnormalized = intrinsics @ point_3d
+        uv = torch.div(uv_unnormalized, uv_unnormalized[..., -1:, :])
+
+        # Do torch.split results in Bx1XN.
+        px, py, _ = torch.split( uv, 1, dim=-2 )
+
+        if normalized:
+            # Using shape - 1 is the way for cv2.remap() and align_corners=True of torch.nn.functional.grid_sample().
+            # px = px / ( self.ss.W - 1 ) * 2 - 1
+            # py = py / ( self.ss.H - 1 ) * 2 - 1
+            # Using shape is the way for torch.nn.functional.grid_sample() with align_corners=False.
+            px = px / self.ss.W * 2 - 1
+            py = py / self.ss.H * 2 - 1
+
+        # pixel_coor = torch.stack( (px, py), dim=0 )
+        pixel_coor = torch.cat( (px, py), dim=-2 )
+
+        # Filter the invalid pixels by the image size. Valid mask takes on shape [B] x N
+        valid_mask_px = torch.logical_and(px < self.ss.shape[0], px > 0)
+        valid_mask_py = torch.logical_and(py < self.ss.shape[1], py > 0)
+        valid_mask = torch.logical_and(valid_mask_py, valid_mask_px)
+
+        # This is for the batched dimension.
+        valid_mask = valid_mask.squeeze(-2)
+
+        return self.out_wrap(pixel_coor), self.out_wrap(valid_mask)
+
+
+    def to_(self, dtype=None, device=None):
+        assert dtype is not None and device is not None, \
+            f'dtype and device cannot both be None. '
+        
+        self.device = device
+        
+    def __deepcopy__(self, memo):

```

Comment:
I think we do not need to override this parent function since Python will figure out the correct value for `self.__class__`. It's going to be the child class.
![deepcopy](https://user-images.githubusercontent.com/20142527/199316420-604a333c-f89b-469d-beeb-f415dda07f2b.png)